tsdb/querier_test.go

```diff
@@ -2194,6 +2195,76 @@ func TestPostingsForMatchers(t *testing.T) {
 	}
 }
 
+// TestQuerierIndexQueriesRace tests the index queries with racing appends.
+func TestQuerierIndexQueriesRace(t *testing.T) {
+	const testRepeats = 1000
+
+	testCases := []struct {
+		matchers []*labels.Matcher
+	}{
+		{
+			matchers: []*labels.Matcher{
+				// This matcher should involve the AllPostings posting list in calculating the posting lists.
+				labels.MustNewMatcher(labels.MatchNotEqual, labels.MetricName, "metric"),
+			},
+		},
+		{
+			matchers: []*labels.Matcher{
+				// The first matcher should be effectively the same as AllPostings, because all series have m=0
+				// If it is evaluated first, then __name__=metric will contain more series than m=0.
+				labels.MustNewMatcher(labels.MatchNotEqual, "m", "0"),
+				labels.MustNewMatcher(labels.MatchEqual, labels.MetricName, "metric"),
+			},
+		},
+	}
+
+	for _, c := range testCases {
+		c := c
+		t.Run(fmt.Sprintf("%v", c.matchers), func(t *testing.T) {
+			db := openTestDB(t, DefaultOptions(), nil)
+			h := db.Head()
+			t.Cleanup(func() {
+				require.NoError(t, db.Close())
+			})
+			ctx, cancel := context.WithCancel(context.Background())
+			wg := &sync.WaitGroup{}
+			wg.Add(1)
+			go appendSeries(t, ctx, wg, h)
+			t.Cleanup(wg.Wait)
+			t.Cleanup(cancel)
+
+			for i := 0; i < testRepeats; i++ {
+				q, err := db.Querier(ctx, math.MinInt64, math.MaxInt64)
+				require.NoError(t, err)
+
+				values, _, err := q.LabelValues("n", c.matchers...)
```

Comment:
Nit: I would be more explicit on the label name, requesting `"this_label_does_not_exist"`, that would save some moments to the reader checking the test 3 times to see that this label isn't defined.

Reply:
Agreed on this comment, it wasn'tt straightforward for me to know what was happening here.

Reply:
In this case `n` always exists, but we expect to get, no values because the matchers select no series. I changed `m` to `always_0` and `n` to `seq` (from "sequence") in c52f29ccf324a11b2d02368105d5e4840e59bedd. Does this make it better?